### ISSUE TITLE
Poor performance of ArcFace loss compared to CrossEntropy

### ISSUE THREAD

**Issue by Mactarvish** [KevinMusgrave/pytorch-metric-learning#351]
I'm trying to use ArcFace to train my model with a face dataset containing 10000 classes (about 3 samples per class). Here's my main training code:
```
import torchvision
import torch
import torch.nn as nn
from pytorch_metric_learning import losses

num_classes = 10000
embedding_size = 128

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.trunk = torchvision.models.resnet18(pretrained=True)
        
        self.trunk_output_size = self.trunk.fc.in_features
        self.trunk.fc = common_functions.Identity()
        self.embedder = torch.nn.Linear(self.trunk_output_size, embedding_size)

    def forward(self, x):
        x = self.trunk(x)
        x = self.embedder(x)
        return x


def train(model, loss_func, device, train_loader, optimizer, W_optimizer, epoch):
    model.train()
    for data, labels in train_loader:
        data, labels = data.to(device), labels.to(device)
        
        optimizer.zero_grad()
        W_optimizer.zero_grad()
        embeddings = model(data)
        loss = loss_func(embeddings, labels)
        loss.backward()
        optimizer.step()
        W_optimizer.step()


if __name__ == "__main__":
    dataset = FaceSet()
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=32, num_workers=32)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = Net().to(device)

    loss_func = losses.ArcFaceLoss(num_classes=num_classes, embedding_size=embedding_size)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=0.0001)
    W_optimizer = torch.optim.Adam(loss_func.parameters(), lr=0.001, weight_decay=0.0001)
    num_epochs = 100

    for epoch in range(1, num_epochs+1):
        train(model, loss_func, device, dataloader.train_loader, optimizer, W_optimizer, epoch)

        save_model_name = './model/model_{}.pth'.format(epoch)
        torch.save(model, save_model_name)
```

I tested the model on a 5000-classes test set, but only got a top10 recall 0.03, while with a normal nn.CrossEntropyLoss the recall is 0.88. Are there any big mistakes in above training code? Or an example code of ArcFace will be appreciated.

**Comment by KevinMusgrave** [comment]
You could try a lower scale. The default is 64.

```python
loss_func = losses.ArcFaceLoss(num_classes=num_classes, embedding_size=embedding_size, scale=16)
```

**Comment by Mactarvish** [comment]
That's the point! I rescaled to 8 and the training merged with final recall 0.91, thanks.
